### PR TITLE
Implemented independent softmax layer & cross-entropy loss 

### PR DESCRIPTION
**Pull request by fufjvnvnf**

### CODE REVIEW

File: mnist_cnn.jl [hunk 1]
```diff
@@ -345,9 +346,8 @@ function Adam(net::SequentialNet, train_set, validation_set;
             append!(all_losses, mean(loss))
             epo_cor  += get_corr(pred, batch_Y)
             local acc = get_corr(pred, batch_Y) / batch_size
-            println("[$(bid)/$(num_batch)] Loss is: $(mean(loss))\tAccuracy:$(acc)")
         end
-
```
Comment (stevenygd) [inline mnist_cnn.jl]: Please remove unrelated lines.

File: mnist_cnn.jl [hunk 2]
```diff
@@ -360,7 +360,8 @@ function Adam(net::SequentialNet, train_set, validation_set;
         end
         append!(val_losses, mean(v_loss))
         append!(val_accu,   mean(v_accu))
-        println("Epo $(epo) has loss :$(mean(v_loss))\t\taccuracy : $(mean(v_accu))")
```
Comment (stevenygd) [inline mnist_cnn.jl]: Same, seems irrelevant to this PR.

File: mnist_cnn_softcross.jl [hunk 3]
```diff
@@ -0,0 +1,119 @@
+include("src/NN.jl")
```
Comment (stevenygd) [inline mnist_cnn_softcross.jl]: I am thinking maybe you should put this into the example directory, or maybe make a benchmark directory to put it. Just don't want to pollute the root directory.

File: src/layers/CrossEntropyLoss.jl [hunk 4]
```diff
@@ -0,0 +1,95 @@
+type CrossEntropyLoss <: LossCriteria
+  x :: Array{Float64}
```
Comment (stevenygd) [inline src/layers/CrossEntropyLoss.jl]: Please add necessary comments about what each variable is planning to store.

File: src/layers/CrossEntropyLoss.jl [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+      out_size = (config["batch_size"], config["input_size"][1])
+  else
+      out_size = getOutputSize(p)
+  end
+  l.classes = out_size[2]
+  l.x      = Array{Float64}(out_size)
+  l.y      = Array{Float64}(out_size)
+end
+
+function convert_to_one_hot(l::CrossEntropyLoss, old_label::Array{Int, 2})
```
Comment (stevenygd) [inline src/layers/CrossEntropyLoss.jl]: I'm thinking maybe you just need to take a vector as input, don't need to care about whether it's one-hot or not. Consider the API for tensor flow: https://www.tensorflow.org/api_docs/python/tf/contrib/losses/softmax_cross_entropy

File: src/layers/CrossEntropyLoss.jl [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+  if size(label)[2] == 1
+    # convert one-dim label to one hot vectors
+    label = convert_to_one_hot(l,label)
+  end
+  # from now on label is guaranteed to be of one-hot
+  @assert size(Y) == size(label)
+  m,n = size(Y)
+  loss = zeros(m)
+  m,n = size(Y)
+  for i=1:m
```
Comment (stevenygd) [inline src/layers/CrossEntropyLoss.jl]: I think we could vectorize this. Maybe consider take a look at the implementation of Lasagne's implementation of cross entropy 
https://github.com/Lasagne/Lasagne/blob/master/lasagne/objectives.py#L147-L174

File: src/layers/SequentialNet.jl [hunk 7]
```diff
@@ -16,7 +16,7 @@ end
  # TODO could remove the local
  # TODO inp could be updated in-places
  # TODO define inplace operator
-function forward(net::SequentialNet, x::Array{Float64}, label::Array; kwargs...)
+function forward(net::SequentialNet, x::Array{Float64}, label::Array{Int,2}; kwargs...)
```
Comment (stevenygd) [inline src/layers/SequentialNet.jl]: Why did you have to do that? I'm thinking about leaving the label as an abstract class, and maybe need to separate out the loss layer in this forward/backward schema.

File: src/layers/SoftMax.jl [hunk 8]
```diff
@@ -1,41 +1,96 @@
...
 
-    for i = 1: ndims(X)
-        t = z[i]
-        w = zeros(ndims(X))
-        w[i] = 1
-        w = w .- z
-        u[:,i] = t * w
+    jacobian = zeros(n,n)
+    for batch=1:m
+      # diagonal
```
Comment (stevenygd) [inline src/layers/SoftMax.jl]: for all the lines that are commented out, if you don't want it delete it please

File: src/layers/SoftMax.jl [hunk 9]
```diff
@@ -1,41 +1,96 @@
...
+        jacobian[i,i] = li*(1-li)
+      end
+      # # n x 1 = n x n * n x 1
+      result[batch,:] = jacobian * DLDY[batch,:]
     end
-    l.last_loss = DLDY' * u
-    return l.last_loss
+
+    return result
+    # sumX = sum(exp(X))
```
Comment (stevenygd) [inline src/layers/SoftMax.jl]: Clean up please.

File: test/SoftMax-CrossEntropyTest.jl [hunk 10]
```diff
@@ -0,0 +1,82 @@
+include("../src/layers/LayerBase.jl")
+include("../src/layers/SoftMaxCrossEntropy.jl")
+include("../src/layers/SoftMax.jl")
+include("../src/layers/CrossEntropyLoss.jl")
+using Base.Test
+
+smaxcross = SoftMaxCrossEntropyLoss()
```
Comment (stevenygd) [inline test/SoftMax-CrossEntropyTest.jl]: Well, I don't think it's safe to use previously implemented layers for testing. Should craft layer specific test cases.

File: test/SoftMax-CrossEntropyTest.jl [hunk 11]
```diff
@@ -0,0 +1,82 @@
+include("../src/layers/LayerBase.jl")
+include("../src/layers/SoftMaxCrossEntropy.jl")
+include("../src/layers/SoftMax.jl")
+include("../src/layers/CrossEntropyLoss.jl")
```
Comment (stevenygd) [inline test/SoftMax-CrossEntropyTest.jl]: Also, please test each layer separately.

File: test/SoftMaxTest.jl [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+end
+
+# First Test
+println("Unit test 1...")
+x = [1.0, 2.0, 3.0]
+y = [0.09003, 0.24473, 0.66524]
+dldy = [1., 1., 1.]
+dldx = [0.00784, 0.00001, 0.00001]
+testSoftMax(x, y, dldy, dldx)
+println("test 1 passed.\n")
```
Comment (stevenygd) [inline test/SoftMaxTest.jl]: This is good, maybe write more tests to cover corner cases.
i.e. numerical stability issues?

File: benchmark/SoftMaxCrossEntropyBenchmark.jl [hunk 13]
```diff
@@ -0,0 +1,63 @@
+include("../src/layers/LayerBase.jl")
+include("../src/layers/SoftMaxCrossEntropy.jl")
+include("../src/layers/SoftMax.jl")
+include("../src/layers/CrossEntropyLoss.jl")
+using Base.Test
+
+smaxcross = SoftMaxCrossEntropyLoss()
```
Comment (stevenygd) [inline benchmark/SoftMaxCrossEntropyBenchmark.jl:7]: The tests shouldn't use the previously implemented layer for testing.

File: benchmark/SoftMaxCrossEntropyBenchmark.jl [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+end
+
+function benchmmark(smaxcross, smax, cross, batch_size, input_size; alpha = 1.)
+  # batch_size = 10
+  # input_size = 5
+  before(smaxcross, smax, cross, Dict{String, Any}("batch_size" => batch_size, "input_size" => [input_size]))
+  x = rand(batch_size, input_size)
+  label = rand(0:input_size-1, batch_size,1)
+  # benchmark for original SoftMaxCrossEntropy
+  tic()
```
Comment (stevenygd) [inline benchmark/SoftMaxCrossEntropyBenchmark.jl]: I would prefer using @time and @times for testing.

File: src/layers/SoftMax.jl [hunk 15]
```diff
@@ -1,41 +1,66 @@
...
+    y :: Array{Float64}
+    has_init :: Bool
+    jacobian :: Array{Float64}
+    dldx :: Array{Float64}
+    # last_loss   :: Array{Float64}
 
     function SoftMax()
-        return new(Float64[], Float64[], Float64[])
+        return new(Float64[], Float64[])
     end
```
Comment (stevenygd) [inline src/layers/SoftMax.jl]: Initialize other parameters as well.

File: src/layers/SoftMax.jl [hunk 16]
```diff
@@ -1,41 +1,66 @@
...
+  m,n = out_size
+  l.jacobian = Array{Float64}(n,n)
+  l.dldx = Array{Float64}(out_size)
+end
+
+function forward(l::SoftMax, X::Array{Float64,2}; kwargs...)
 
+    l.x = X
+    # iterating each row/picture
+    lxexp = exp(l.x)
```
Comment (stevenygd) [inline src/layers/SoftMax.jl]: You could replace these two lines as in-place operations.

File: src/layers/SoftMaxCrossEntropy.jl [hunk 17]
```diff
@@ -52,6 +52,7 @@ function update(l::SoftMaxCrossEntropyLoss, input_size::Tuple;)
 end
 
 function forward(l::SoftMaxCrossEntropyLoss, Y::Array{Float64,2}, label::Array{Int, 2}; kwargs...)
+    # TODO this is a questionable invariant; [label] ranges from 0 to 9 instead of being one hot
```
Comment (stevenygd) [inline src/layers/SoftMaxCrossEntropy.jl:55]: I don't think this is valid : (.

**Review by stevenygd — COMMENTED** [review]
Please response to these comments.

File: mnist_cnn_softcross.jl (hunk 3 above)
Comment (fufjvnvnf) [reply mnist_cnn_softcross.jl]: resolved by putting into `examples` dir

File: src/layers/SoftMaxCrossEntropy.jl (hunk 17 above)
Comment (fufjvnvnf) [reply src/layers/SoftMaxCrossEntropy.jl:55]: :(

File: src/layers/SoftMax.jl (hunk 15 above)
Comment (fufjvnvnf) [reply src/layers/SoftMax.jl]: Resolved.

File: src/layers/SequentialNet.jl (hunk 7 above)
Comment (fufjvnvnf) [reply src/layers/SequentialNet.jl]: Resolved by changing back.

File: src/layers/SoftMax.jl (hunk 8 above)
Comment (fufjvnvnf) [reply src/layers/SoftMax.jl]: Resolved.

File: src/layers/CrossEntropyLoss.jl (hunk 4 above)
Comment (fufjvnvnf) [reply src/layers/CrossEntropyLoss.jl]: Resolved.

File: src/layers/CrossEntropyLoss.jl (hunk 5 above)
Comment (fufjvnvnf) [reply src/layers/CrossEntropyLoss.jl]: Resolved.

File: benchmark/SoftMaxCrossEntropyBenchmark.jl (hunk 13 above)
Comment (stevenygd) [reply benchmark/SoftMaxCrossEntropyBenchmark.jl:7]: Please don't use previously implemented layers to test new layers. Write separate tests and keep things decouple.

File: benchmark/SoftMaxCrossEntropyBenchmark.jl (hunk 14 above)
Comment (stevenygd) [reply benchmark/SoftMaxCrossEntropyBenchmark.jl]: I remember there are reasons preferring @time to tic() toc()s. Please make the change.

File: examples/mnist_cnn_softcross.jl [hunk 18]
```diff
@@ -0,0 +1,119 @@
...
+epo_losses, epo_accu, val_losses, val_accu, all_losses = Adam(
+    net, (trX, trY), (valX, valY);
+    ttl_epo = 10, batch_size = batch_size,
+    lrSchedule = x -> 0.001, verbose=1
+)
+
+figure(figsize=(12,6))
+# subplot(221)
+plot(1:length(all_losses), all_losses)
+title("Nestorove Momemtum : Training losses")
```
Comment (stevenygd) [inline examples/mnist_cnn_softcross.jl]: I thought you use Adam. Not nestorove Momentum.

File: src/layers/CrossEntropyLoss.jl [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+
+"""
+Invaraint for this to work:
+  Label needs to be either a vector specifying each item's class or a matrix made up with multiple one hot vectors.
+  In the former case, labels need to be in the range of 0:classes-1.
+"""
+function forward(l::CrossEntropyLoss, Y::Array{Float64,2}, label::Array{Float64, 2}; kwargs...)
+  if size(label)[2] == 1
+    # convert one-dim label to one hot vectors
+    label = convert_to_one_hot(l,label)
```
Comment (stevenygd) [inline src/layers/CrossEntropyLoss.jl]: Why do you still need `convert_to_one_hot` function? Based on your design it shouldn't be needed.

File: src/layers/CrossEntropyLoss.jl [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+
+function backward(l::CrossEntropyLoss, label::Array{Float64, 2};kwargs...)
+  if size(label)[2] == 1
+    # convert one-dim label to one hot vectors
+    label = convert_to_one_hot(l,label)
+  end
+
+  @assert size(l.x) == size(label)
+
+  l.dldx = -label./l.x
```
Comment (stevenygd) [inline src/layers/CrossEntropyLoss.jl]: I'm curious, have you profile this line? It occurs to me that this line might have additional allocation. But obviously there is nothing needed to be allocated.

File: mnist_cnn.jl (hunk 2 above)
Comment (fufjvnvnf) [reply mnist_cnn.jl]: Resolved

File: src/layers/SoftMax.jl [hunk 21]
```diff
@@ -1,41 +1,56 @@
...
+  m,n = out_size
+  l.jacobian = Array{Float64}(n,n)
+  l.dldx = Array{Float64}(out_size)
+end
+
+function forward(l::SoftMax, X::Array{Float64,2}; kwargs...)
+
+    l.x = X
+    # iterating each row/picture
+    lxexp = exp(l.x)
```
Comment (stevenygd) [inline src/layers/SoftMax.jl]: You got the idea, but I'm still curious, how much allocation have you reduce by writing it this way? Does it make it in-place?

File: src/layers/SoftMax.jl [hunk 22]
```diff
@@ -1,41 +1,56 @@
...
+    # credits: https://stats.stackexchange.com/questions/79454/softmax-layer-in-a-neural-network?newreg=d1e89b443dd346ae8bccaf038a944221
+    m,n =size(l.x)
 
-    for i = 1: ndims(X)
-        t = z[i]
-        w = zeros(ndims(X))
-        w[i] = 1
-        w = w .- z
-        u[:,i] = t * w
+    ly = Array{Float64}(n)
```
Comment (stevenygd) [inline src/layers/SoftMax.jl]: Why do you have to allocate this every function  call? `n` is fixed once the layer is initialized.

File: src/layers/SoftMaxCrossEntropy.jl (hunk 17 above)
Comment (stevenygd) [reply src/layers/SoftMaxCrossEntropy.jl:55]: I meant this is wrong, and you should make the label into Float array.

**Review by stevenygd — COMMENTED** [review]
More to go.

File: examples/mnist_cnn_softcross.jl (hunk 18 above)
Comment (fufjvnvnf) [reply examples/mnist_cnn_softcross.jl]: Yea, I forgot to change the title; so is mnist_cnn.jl

File: src/layers/SoftMax.jl (hunk 21 above)
Comment (fufjvnvnf) [reply src/layers/SoftMax.jl]: I never profiled this line but this should make the operation in-place. Forward has always been fast so I didn't noted the changes to speed by writing in this way, but forward now for the the layers combined takes slightly less time than the old layer.

File: test/SoftMax-CrossEntropyTest.jl (hunk 11 above)
Comment (fufjvnvnf) [reply test/SoftMax-CrossEntropyTest.jl]: Resolved by creating "CrossEntropyTest.jl"; keeping "SoftMax-CrossEntropyTest.jl" just in case.

File: test/SoftMax-CrossEntropyTest.jl (hunk 10 above)
Comment (fufjvnvnf) [reply test/SoftMax-CrossEntropyTest.jl]: Resolved.

File: benchmark/SoftMaxCrossEntropyBenchmark.jl (hunk 13 above)
Comment (fufjvnvnf) [reply benchmark/SoftMaxCrossEntropyBenchmark.jl:7]: Resolved in `CrossEntropyTest.jl` and `SoftMaxTest.jl`

File: benchmark/SoftMaxCrossEntropyBenchmark.jl (hunk 14 above)
Comment (fufjvnvnf) [reply benchmark/SoftMaxCrossEntropyBenchmark.jl]: I will stick with using`tic() toq()` here since according to documentation  https://docs.julialang.org/en/release-0.4/manual/performance-tips/, "... This is the single biggest advantage of @time vs. functions like tic() and toc(), which only report time"; these two functions are virtually the same but I'll stick with `tic()` here since I would only want time to be reported here since this is a benchmark test in which I want merely a broad sense of time spent by each function; I usually use @time when inspecting into details of methods to figure out whether it has given too many allocations.

File: src/layers/SoftMaxCrossEntropy.jl (hunk 17 above)
Comment (fufjvnvnf) [reply src/layers/SoftMaxCrossEntropy.jl:55]: Resolved by switching to `Float64`.

File: src/layers/CrossEntropyLoss.jl (hunk 19 above)
Comment (fufjvnvnf) [reply src/layers/CrossEntropyLoss.jl]: Planning to resolve by checking if label is a matrix or vector, and conduct `convert_to_one_hot ` accordingly in `SequentialNet` layer. What do you think?

File: src/layers/CrossEntropyLoss.jl (hunk 20 above)
Comment (fufjvnvnf) [reply src/layers/CrossEntropyLoss.jl]: What do you mean by "additional allocation"? If it because I'm doing `-` at the same time thus the program would look for somewhere to allocate the result of negation first?

File: src/layers/SoftMax.jl (hunk 22 above)
Comment (fufjvnvnf) [reply src/layers/SoftMax.jl]: `ly` should serve as a cache to hold array of length `n` during the construction of `jacobian` at each `batch`. Resolved by adding `l.ly` as `Softmax`'s instance and allocate it only once during `init`.

File: examples/mnist_cnn_softcross.jl (hunk 18 above)
Comment (stevenygd) [reply examples/mnist_cnn_softcross.jl]: Then change it, make it less confusing.

File: src/layers/CrossEntropyLoss.jl (hunk 19 above)
Comment (stevenygd) [reply src/layers/CrossEntropyLoss.jl]: I don't think we want to do convert_to_one_hot anywhere in our framework. The clients should do that.

File: src/layers/CrossEntropyLoss.jl (hunk 20 above)
Comment (stevenygd) [reply src/layers/CrossEntropyLoss.jl]: Try @time with it and see how the allocations changes according to array size.

File: src/layers/SoftMax.jl (hunk 21 above)
Comment (stevenygd) [reply src/layers/SoftMax.jl]: Time profiling isn't a good measurement and it's quite unreliable, depends on how many programs you run and how OS schedules the job. So being faster in one specific bench mark doesn't mean that it will be faster in overall performance. I would prefer looking at locations since there is no doubt that the more you allocate the slower you will be. Once the allocations is in the minimal level, then we could discuss some other optimizations.

File: src/layers/SoftMax.jl (hunk 22 above)
Comment (stevenygd) [reply src/layers/SoftMax.jl]: Well, it's a good plan, I would prefer seeing the changes first.

File: src/layers/SoftMaxCrossEntropy.jl (hunk 17 above)
Comment (stevenygd) [reply src/layers/SoftMaxCrossEntropy.jl:55]: I didn't see the change.

File: test/SoftMax-CrossEntropyTest.jl (hunk 11 above)
Comment (stevenygd) [reply test/SoftMax-CrossEntropyTest.jl]: You can keep it in your local git, don't push it to the repo.

File: test/SoftMax-CrossEntropyTest.jl (hunk 10 above)
Comment (stevenygd) [reply test/SoftMax-CrossEntropyTest.jl]: But I didn't see the change.

File: src/layers/CrossEntropyLoss.jl (hunk 19 above)
Comment (fufjvnvnf) [reply src/layers/CrossEntropyLoss.jl]: Labels are generated by the `datasplit` method under `dataset.jl`. Should I alter that method to ensure matrix labels be returned?

File: test/SoftMax-CrossEntropyTest.jl (hunk 10 above)
Comment (fufjvnvnf) [reply test/SoftMax-CrossEntropyTest.jl]: Resolved by creating separate tests for `SoftMax` and `CrossEntropyLoss` as mentioned in other reviews; `SoftMax-CrossEntropyTest.jl` is technically not needed and will be deleted in next commit unless otherwise requested.

File: test/SoftMax-CrossEntropyTest.jl (hunk 11 above)
Comment (fufjvnvnf) [reply test/SoftMax-CrossEntropyTest.jl]: Roger.

File: src/layers/SoftMaxCrossEntropy.jl (hunk 17 above)
Comment (fufjvnvnf) [reply src/layers/SoftMaxCrossEntropy.jl:55]: Under the new interface, `SoftMaxCrossEntropy` behaves virtually the same as the two separate newly implemented layers. Should I just delete this old layer?

File: examples/mnist_cnn_softcross.jl (hunk 18 above)
Comment (fufjvnvnf) [reply examples/mnist_cnn_softcross.jl]: Resolved in both mains.

File: src/layers/CrossEntropyLoss.jl (hunk 19 above)
Comment (fufjvnvnf) [reply src/layers/CrossEntropyLoss.jl]: Still awaiting for clarification on this subject.

File: src/layers/CrossEntropyLoss.jl (hunk 20 above)
Comment (fufjvnvnf) [reply src/layers/CrossEntropyLoss.jl]: Reimplemented with a one-lined plausibly minimal allocation version.

File: src/layers/SoftMax.jl (hunk 21 above)
Comment (fufjvnvnf) [reply src/layers/SoftMax.jl]: Both `Softmax` and `CrossEntropy` are now reimplemented in the most concise fashion with preallocations also been taken care. Dot operations though, i.e. dot division in `SoftMax.jl` line 39 and dot multiplication `CrossEntropyLoss.jl` line 36, takes unreasonable allocations even though I preallocate their output, especially for the latter, where the negation `-` takes same amount of allocations as the `.*`. Looking for a more efficient way here.

File: src/layers/SoftMax.jl (hunk 22 above)
Comment (fufjvnvnf) [reply src/layers/SoftMax.jl]: Reimplemented `SoftMax` forward

File: src/layers/SoftMax.jl (hunk 16 above)
Comment (fufjvnvnf) [reply src/layers/SoftMax.jl]: Reimplemented `SoftMax` forward as mention in other replies. Not sure if they are `in-place` as you expect.

File: test/SoftMaxTest.jl (hunk 12 above)
Comment (fufjvnvnf) [reply test/SoftMaxTest.jl]: Resolved by writing tests with extreme values. Tests cases could only go so far as numbers around the rand of 700~ since exponential of numbers higher than that result in `inf` in `julia`.

File: src/layers/SoftMax.jl (hunk 9 above)
Comment (fufjvnvnf) [reply src/layers/SoftMax.jl]: Resolved.

File: src/layers/CrossEntropyLoss.jl (hunk 6 above)
Comment (fufjvnvnf) [reply src/layers/CrossEntropyLoss.jl]: Resolved by reimplemented `CrossEntropyLoss` forward.

File: mnist_cnn.jl (hunk 1 above)
Comment (fufjvnvnf) [reply mnist_cnn.jl]: Resolved.

**Comment by fufjvnvnf** [comment]
1. Should `label` be converted to matrices in `dataSplit` or in mains?
2. `Forward` in `SoftMax` and `CrossEntropyLoss` are reimplemented. Please check to see if minimal allocation has been achieved.
3. Under the new loss function interface, original `SoftMaxCrossEntropy` needs to be complete reimplemented since it assumes a vector label, in which case however the code would be virtually the same with the newly implemented separate layers. Shouldn't we just delete this old layer?

**Comment by linzhiqiu** [comment]
你好棒棒哦

File: src/layers/SoftMaxCrossEntropy.jl (hunk 17 above)
Comment (stevenygd) [reply src/layers/SoftMaxCrossEntropy.jl:55]: No, keep it, I think it will be useful later on.

File: benchmark/SoftMaxCrossEntropyBenchmark.jl (hunk 14 above)
Comment (stevenygd) [reply benchmark/SoftMaxCrossEntropyBenchmark.jl]: I prefer a benchmark that could clearly tell me what's the execution time and allocations. Please check @times for it.

File: mnist_cnn_softcross.jl (hunk 3 above)
Comment (stevenygd) [reply mnist_cnn_softcross.jl]: Then you should probably checkout this file.

File: benchmark/SoftMaxCrossEntropyBenchmark.jl [hunk 23]
```diff
@@ -0,0 +1,70 @@
...
+
+function benchmmark(smaxcross, smax, cross, batch_size, input_size; alpha = 1.)
+  # batch_size = 10
+  # input_size = 5
+  before(smaxcross, smax, cross, Dict{String, Any}("batch_size" => batch_size, "input_size" => [input_size]))
+  x = rand(batch_size, input_size)
+  label = rand(0:input_size-1, batch_size,1)
+  onehot_label = convert_to_one_hot(label,input_size)
+  # benchmark for original SoftMaxCrossEntropy
+  tic()
```
Comment (stevenygd) [inline benchmark/SoftMaxCrossEntropyBenchmark.jl:34]: I still think that as a benchmark you should be able to give both the execution time and memory allocation, since both will be metrics to be benchmarked. In this case, `@times` will be a better option, since it could be used to accomplish both.

**Review by stevenygd — COMMENTED** [review]
Just couple more.

**Comment by stevenygd** [comment]
@fufjvnvnf 
1. Label should be a matrix with one-hot row vectors when being output from the `datasets.jl`. You could create an issue and make this as a TODO.
2. Will check it soon.
3. I don't think `SoftmaxCrossEntropyLoss` needs to be reimplemented. It should be easily adapted.

File: benchmark/SoftMaxCrossEntropyBenchmark.jl [hunk 24]
```diff
@@ -0,0 +1,70 @@
...
+  # d2 = @time backward(smax, backward(cross, label))
+  d2 = backward(smax, backward(cross, onehot_label))
+
+  time = toq()
+  println("new backward uses: ", time)
+
+  println("Done")
+end
+
+for i=1:20
```
Comment (stevenygd) [inline benchmark/SoftMaxCrossEntropyBenchmark.jl:68]: I think what you want is to run 20 iterations and average the results, so that the result will be more reliable. Otherwise there is not much difference between running 2 times and running 20 times.

File: benchmark/SoftMaxCrossEntropyBenchmark.jl [hunk 25]
```diff
@@ -0,0 +1,70 @@
...
+
+function convert_to_one_hot(x::Array{Int64}, classes)
+  m = zeros(size(x,1), classes)
+  for i=1:size(x,1)
+    m[i,x[i]+1]=1
+  end
+  m
+end
+
+function benchmmark(smaxcross, smax, cross, batch_size, input_size; alpha = 1.)
```
Comment (stevenygd) [inline benchmark/SoftMaxCrossEntropyBenchmark.jl]: What's the `alpha` for?

Comment (fufjvnvnf) [reply benchmark/SoftMaxCrossEntropyBenchmark.jl]: Does not seem to be of any use; resolved by removing.

**Comment by fufjvnvnf** [comment]
@stevenygd How should we allow users to retrieve the network's prediction result? Shouldn't it make more sense to keep `pred` as an output for loss functions?

**Comment by stevenygd** [comment]
@fufjvnvnf Consider the interface for Tensorflow.

**Comment by stevenygd** [comment]
I will merge it first and put the rest works for next iteration. Future works include:
1. Improve softmax and CrossEntropy backward efficiency
2. Make more reliable benchmark statistics and benchmark set.

File: src/layers/SoftMaxCrossEntropy.jl (hunk 17 above)
Comment (fufjvnvnf) [reply src/layers/SoftMaxCrossEntropy.jl:55]: Resolved by bring `SoftMaxCrossEntropy` to its original state; going to resolve this in next pr
